katdal/applycal.py

```diff
@@ -0,0 +1,188 @@
+    cache : :class:`katdal.sensordata.SensorCache` object
+        Sensor cache, used to look up individual correction sensors
+    inputs : sequence of string
+        Correlator input labels
+    input1_index, input2_index : list of int, length n_corrprods
+        Indices into `inputs` of first and second items of correlation product
+    cal_products : sequence of string
+        Calibration products that will contribute to corrections
+
+    Returns
+    -------
+    gains : array of complex64, shape (n_chans, n_corrprods)
+        Gain corrections per channel per correlation product
+
+    Raises
+    ------
+    KeyError
+        If input and/or cal product has no associated correction
+    """
+    g_per_input = np.ones((len(inputs), len(channels)), dtype='complex64')
+    for product in cal_products:
+        for n, inp in enumerate(inputs):
+            sensor_name = 'Calibration/{}_correction_{}'.format(inp, product)
+            g_product = cache.get(sensor_name)[dump]
+            if np.shape(g_product) != ():
+                g_product = g_product[channels]
+            g_per_input[n] *= g_product
+    g_per_cp = g_per_input[input1_index] * g_per_input[input2_index].conj()
+    return g_per_cp.T
+
+
+def apply_vis_correction(out, correction):
+    """Apply `correction` in-place to visibility data in `out`."""
+    correction[np.isnan(correction)] = np.complex64(1)
+    out *= correction
+
+
+def add_applycal_transform(indexer, cache, corrprods, cal_products,
+                           apply_correction):
+    """Add transform to indexer that applies calibration corrections.
+
+    This adds a transform to the indexer which wraps the underlying data
+    (visibilities, weights or flags). The transform will apply all calibration
+    corrections specified in `cal_products` to each dask chunk individually.
+    The actual application method is also user-specified, which allows most
+    of the machinery to be reused between visibilities, weights and flags.
+    The time and frequency selections are salvaged from `indexer` but the
+    selected `corrprods` still needs to be passed in as a parameter to identify
+    the relevant inputs in order to access correction sensors.
+
+    Parameters
+    ----------
+    indexer : :class:`katdal.lazy_indexer.DaskLazyIndexer` object
+        Indexer with underlying dask array that will be transformed
+    cache : :class:`katdal.sensordata.SensorCache` object
+        Sensor cache, used to look up individual correction sensors
+    corrprods : sequence of (string, string)
+        Selected correlation products as pairs of correlator input labels
+    cal_products : sequence of string
+        Calibration products that will contribute to corrections
+    apply_correction : function, signature ``out = f(out, correction)``
+        Function that will actually apply correction to data from indexer
+    """
+    if indexer.dataset.chunks[2] != (len(corrprods),):
```

Comment:
Accessing `dataset` here will trigger the creation of the graph. This work should probably be done inside the transform function so that it is only run when the user triggers creation of the graph.

Reply:
Right, will do.